text/0000-incremental-compilation.md

```diff
@@ -0,0 +1,339 @@
+
+### Global Switches Influencing Codegen
+There are many compiler flags that change the way the generated code looks like, e.g. optimization and debuginfo levels. A simple strategy to deal with this would be to store the set of compiler flags used for building the cache and clearing the cache completely if another set of flags is used. Another option is to keep multiple caches, each for a different set of compiler flags (e.g. keeping both on disk, a 'debug build cache' and a 'release build cache').
+
+### Source Location Encoding
+The current `codemap` implementation in `rustc` stores everything in one global 'address space'. This is unsuited for incremental compilation since old versions of source files need to be removed from the codemap and new source files need to be added. This leads to a similar problem as the sequentially assigned AST node IDs.
+
+It would be better to store span information in byte-offsets relative to a given source file. Then modifying the codemap is possible without invalidating existing source locations stored somewhere in the cache.
+
+### Physical Cache Structure
+The cache could be kept somewhere within the output directory. Object code could be maintained in an `ar` archive so the linker can directly access it. Something similar might be possible for LLVM bitcode.
+
+### Automatic inter-function optimizations
+It should not be too hard to let the compiler keep track of which parts of the program change infrequently and then let it speculatively build object files with more than one function in them. For these aggregate object files inter-function LLVM optimizations could then be enabled, yielding faster object code at little additional cost. Other strategies for controlling cache granularity can be implemented in a similar fashion.
+
+### Parallelization
+If some care is taken in implementing the above concepts it should be rather easy to do translation and codegen in parallel for all items, since by design we already have (or can deterministically compute) all the information we need.
+
+# Drawbacks
+
+Implementing this will need a lot of work on the compiler architecture. But that's something that will be needed sooner or later anyway.
+
+# Alternatives
+
+I'd definitely like to hear about them.
+
+# Unresolved questions
+
+## Dependency Graph Construction before Type Inference
+I'm not sure if it would be possible to construct valid dependency graphs *before* type-inference or if that would miss some dependency edges. Or more generally, how much per-item work can be pushed until after caching strikes.
```

Comment:
Thanks for the comment, @epdtry! 

It occurred to me that the `node template` graph, together with the set of visible traits forms the dependency graph for type inference. So, while it may not be possible to accurately compute the dependencies of a function body without doing type inference, it should be possible to cache type inference results just like other `compilation artefacts`.